Issue/2756 update downloadable file info for a product - part 1

partho-maple opened the pull request:
Fixes: #2756 
Part 2: https://github.com/woocommerce/woocommerce-ios/pull/2877

## Description
This PR allows updating the downloadable files info with their details. This PR implements the following capabilities.
- Edit individual downloadable files info in a new scene, when tapped from the downloadable file list table view cell

## Changes
- Add `ProductDownloadFileViewController` and related files to show individual downloadable files' details to edit when tappen from `ProductDownloadListViewController`s' `tableView` cell 
- Encodes downloadable files related properties in Product.swift file
- dropSessionDidUpdate method updated to handle datasource modification
- Handle validation on product update

## Test Cases
Test cases have been included with this PR

## Screenshots

<img width="584" alt="Screen Shot 2020-09-25 at 1 51 12 AM" src="https://user-images.githubusercontent.com/1577598/94197984-5e7f1800-fed8-11ea-89e3-dd6c7b3e8ad8.png">
<img width="584" alt="Screen Shot 2020-09-25 at 1 51 38 AM" src="https://user-images.githubusercontent.com/1577598/94198007-66d75300-fed8-11ea-8d09-d0df0a78f1e0.png">


## Testing

#### Prerequisites
- Create a downloadable product from wp-admin
- Add a few downloadable files from there
- Then navigate to the downloadable files' list scene on the app

#### Case 1 
- Tap on any downloadable file's cell from the list which will navigate to the details scene.
- Check the scene navigation bar title is properly updated with the filename
- Write an invalid URL
- Notice the `Done` button is disabled
- Add a different valid URL
- Notice the `Done` button is enabled
- Now add the previous original URL, and notice the `Done` button is disabled again

#### Case 2
- Do the same changes as `Case 1`, but this time for Filename.
- `Done` button should be disabled when the file name is empty or the same as the original name.
- Notice the downloadable files count to reflect the changes

#### Case 3
- Go to the same screen like the above 2 cases and make some valid changes, press done
- Navigate back to the product details screen and press update.
- Verify that the downloadable file information had been properly updated.
- Also verify from the wp-admin portal

#### Case 4
- Navigate to the "Downloadable Files List" scene and reorder files, press `Done`
- Navigate back to the product details screen and press update.
- Verify that the downloadable file information had been properly updated.
- Also verify from the wp-admin portal

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

partho-maple committed:
downloadable files related properties encoded in Product.swift file

partho-maple committed:
dropSessionDidUpdate method updated to handle datasource modification and bug fix

partho-maple committed:
append and update methods added to ProductDownloadListViewModelOutput

partho-maple committed:
ProductFormViewModel update to hold downloadable files related data for updated product

partho-maple committed:
ProductDownloadSettingsViewController and ProductDownloadFileViewController scene added with complete ui logic implementation

partho-maple committed:
update ProductDownloadListViewController to handle settings and add-edit downloadable file scene

partho-maple committed:
unused codes cleanup

partho-maple committed:
Merge branch 'issue/2755-display-downloadable-files-for-a-product-test-cases' of github.com:partho-maple/woocommerce-ios into issue/2756-update-downloadable-file-info-for-a-product

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

partho-maple committed:
reapplying the changes on project file, that have been reverted due to previous merge conflict

partho-maple committed:
unit test case added and download settings related files removed

partho-maple committed:
Merge branch 'feat/products-downloadable-files' of github.com:woocommerce/woocommerce-ios into issue/2756-update-downloadable-file-info-for-a-product

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj
#	Resolved using theirs, and incorporated my changes appropriately

partho-maple committed:
test case renamed on snake_case

partho-maple committed:
delete button added on ProductDownloadFileViewController

pmusolino commented on WooCommerce/WooCommerce.xcodeproj/project.pbxproj:
```diff
@@ -3256,9 +3266,21 @@
+				773077EA251E943600178696 /* ProductDownloadFileViewController.swift */,
+				773077EB251E943600178696 /* ProductDownloadFileViewController.xib */,
+			);
+			path = "Add or Edit FIle";
```
The name of the folder contains a typo. Should be `Add or Edit File`

partho-maple replied:
😛 fixed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+import UIKit
+import Yosemite
+
+// MARK: - ProductDownloadFileViewController
```
I don't think that this description is really useful. You should remove the `MARK:` and add maybe a small description of what this view controller displays.

partho-maple replied:
RIght. updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+//
+final class ProductDownloadFileViewController: UIViewController {
+
+    @IBOutlet weak var tableView: UITableView!
```
nit: `private`

partho-maple replied:
Updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    private let viewModel: ProductDownloadFileViewModelOutput & ProductDownloadFileActionHandler
+    private var sections: [Section] = []
+    private var error: String?
```
nit: remove it, since it's not used
```suggestion
```

partho-maple replied:
Removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+    //
+    typealias Completion = (_ fileName: String?,
+        _ fileURL: String?,
+        _ fileID: String?,
```
Is possible to not fulfill these info? Maybe the `fileName` and the `fileID`, but why the URL?

partho-maple replied:
I was not sure about the behavior at that time. You are right. URL should not be optional. updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+            self?.handleKeyboardFrameUpdate(keyboardFrame: keyboardFrame)
+        }
+        return keyboardFrameObserver
+    }()
```
nit: 
```suggestion
    private lazy var keyboardFrameObserver = KeyboardFrameObserver { [weak self] keyboardFrame in
        self?.handleKeyboardFrameUpdate(keyboardFrame: keyboardFrame)
    }
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    /// Init
+    ///
+    init(product: ProductFormDataModel, downloadFileIndex: Int = -1, formType: FormType, completion: @escaping Completion) {
```
It's not clear to me why pass `-1` as a default value of `downloadFileIndex`.

partho-maple replied:
Hmm. it's better to make it optional I think.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+    private lazy var throttler: Throttler = Throttler(seconds: 0.5)
+
+    init(product: ProductFormDataModel,
+         downloadFileIndex: Int = -1,
```
Also here, It's not clear to me why pass `-1` as a default value of `downloadFileIndex`.

partho-maple replied:
made it optional

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+extension ProductDownloadFileViewController {
+
+    struct Section: RowIterable, Equatable {
+        let errorTitle: String?
```
Do we use `errorTitle`? It seems unused

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        let footer: String?
+        let rows: [Row]
+
+        init(errorTitle: String? = nil, footer: String? = nil, rows: [Row]) {
```
Same here about `errorTitle`

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        registerTableViewHeaderFooters()
+    }
+
+    /// Since there is only a text field in this view for Product SKU form, the text field becomes the first responder immediately when the view did appear
```
nit: this is wrong 😅

partho-maple replied:
😝 Changed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    /// Since there is only a text field in this view for Product SKU form, the text field becomes the first responder immediately when the view did appear
+    ///
+    func configureTextFieldAsFirstResponder() {
```
nit: which text field? This takes the first available text field in the list. In this list we have two textfields, so it would be better to specify or use another method.

partho-maple replied:
😝 Changed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        for kind in headersAndFooters {
+            tableView.register(kind.loadNib(), forHeaderFooterViewReuseIdentifier: kind.reuseIdentifier)
+        }
+    }
```
We don't show any error section in this view. I think that you copy-pasted it wrongly from the Product SKU VC 😅 Am I wrong?

partho-maple replied:
😝 you are right.  Updated 😅

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    @objc private func deleteDownloadableFile() {
+
```
nit: add `TODO:`

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+    @IBOutlet weak var tableView: UITableView!
+
+    private let viewModel: ProductDownloadFileViewModelOutput & ProductDownloadFileActionHandler
+    private var sections: [Section] = []
```
It's not clear to me why you declared a new `sections` array, since the same array it's available in the `viewModel`.

partho-maple replied:
I used it for initially but forgot to remove 😛

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    func configureSections() {
+        sections = viewModel.sections
+    }
```
nit: this is not needed if we use `viewModel` directly.

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+            return nil
+        }
+        return tableView.cellForRow(at: indexPath) as? TitleAndTextFieldTableViewCell
+    }
```
nit: I think that these two methods are not needed.

partho-maple replied:
cleaned up

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    func rowAtIndexPath(_ indexPath: IndexPath) -> Row {
+        return sections[indexPath.section].rows[indexPath.row]
+    }
```
nit: this is probably not needed if we use it only inside `cellForRowAt`

partho-maple replied:
Cleaned up

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        let cellViewModel = Product.createDownloadFileUrlViewModel(fileUrl: viewModel.fileURL) { [weak self] value in
+            self?.viewModel.handleFileUrlChange(value) { [weak self] (isValid, shouldBringUpKeyboard) in
+                self?.enableDoneButton(isValid)
+                if shouldBringUpKeyboard {
```
nit: also here I think that `shouldBringUpKeyboard` can be removed

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        let cellViewModel = Product.createDownloadFileNameViewModel(fileName: viewModel.fileName) { [weak self] value in
+            self?.viewModel.handleFileNameChange(value) { [weak self] (isValid, shouldBringUpKeyboard) in
+                self?.enableDoneButton(isValid)
+                if shouldBringUpKeyboard {
```
nit: I think that `shouldBringUpKeyboard` can be removed

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+            self?.viewModel.handleFileUrlChange(value) { [weak self] (isValid, shouldBringUpKeyboard) in
+                self?.enableDoneButton(isValid)
+                if shouldBringUpKeyboard {
+                    self?.getFileUrlCell()?.textFieldBecomeFirstResponder()
```
```suggestion
                    cell.textFieldBecomeFirstResponder()
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+            self?.viewModel.handleFileNameChange(value) { [weak self] (isValid, shouldBringUpKeyboard) in
+                self?.enableDoneButton(isValid)
+                if shouldBringUpKeyboard {
+                    self?.getFileNameCell()?.textFieldBecomeFirstResponder()
```
```suggestion
                    cell.textFieldBecomeFirstResponder()
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+private extension ProductDownloadFileViewController {
+
+    func configureNavigationBar() {
+        title = NSLocalizedString(viewModel.formType == .add ? "Add Downloadable File" : viewModel.fileName ?? "",
```
nit: we never use conditional scopes inside an `NSLocalizedString`, also because this probably doesn't work properly with our Localization script.

partho-maple replied:
Sorry, I wasn't aware of that. Updated the code

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    func configureNavigationBar() {
+        title = NSLocalizedString(viewModel.formType == .add ? "Add Downloadable File" : viewModel.fileName ?? "",
+                                  comment: "Individual downloadable file navigation title")
```
Nit
```suggestion
                                  comment: "Downloadable file screen navigation title")
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+                                                          comment: "The action to delete downloadable file for the list")
+            return button
+        }()
+        rightBarButtonItems.append(deleteDownloadableFileBarButton)
```
nit: this is not the button for deleting a downloadable file, but the button for presenting the action sheet that contains the delete button. 

`let moreBarButton`

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+        let updateButtonTitle = NSLocalizedString("Update",
+                                                comment: "Action for updating a Products' downloadable files' info remotely")
+        let UpdateBarButton = UIBarButtonItem(title: updateButtonTitle,
```
```suggestion
        let updateBarButton = UIBarButtonItem(title: updateButtonTitle,
```

partho-maple replied:
typo 😛 
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        navigationItem.rightBarButtonItems = rightBarButtonItems
+
+        removeNavigationBackBarButtonText()
+        self.enableDoneButton(false)
```
nit: `self` is not needed.

partho-maple replied:
yeap

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+                                         target: self,
+                                         action: #selector(deleteDownloadableFile))
+            button.accessibilityTraits = .button
+            button.accessibilityLabel = NSLocalizedString("Delete downloadable file from list",
```
The `accessibilityLabel` description is wrong since this is not the delete button, but I'm also thinking why this is the only `UIBarButton` with `accessibilityLabel`.

partho-maple replied:
Updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+                                         style: .plain,
+                                         target: self,
+                                         action: #selector(deleteDownloadableFile))
+            button.accessibilityTraits = .button
```
nit: not needed, it's already recognized as a button.

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+        view.addSubview(tableView)
+        tableView.translatesAutoresizingMaskIntoConstraints = false
+        view.pinSubviewToSafeArea(tableView)
```
😨 uhm... why adding the tableview as a subview, since it was already added in the xib?

partho-maple replied:
right. removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    private func enableDoneButton(_ enabled: Bool) {
```
nit: `private` is not needed since the extension is already declared `private`.

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+    }
+
+    private func enableDoneButton(_ enabled: Bool) {
+        navigationItem.rightBarButtonItems?.last?.isEnabled = enabled
```
This is not totally safe. What happens if the order of the bar buttons changes? We risk to enable or disable the wrong button.

partho-maple replied:
Updated. Please check if it's OK now.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+//
+private extension ProductDownloadFileViewController {
+
+    /// Displays a Notice onscreen, indicating that you can't add a sale price without adding before the regular price
```
😅 the comment here is wrong

partho-maple replied:
updated 😝

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+    ///
+    func displayEmptyFileNameErrorNotice() {
+        UIApplication.shared.keyWindow?.endEditing(true)
+        let message = NSLocalizedString("File name can not be empty",
```
nit:
```suggestion
        let message = NSLocalizedString("The file name can not be empty",
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    /// Displays a Notice onscreen, indicating that you can't add a sale price without adding before the regular price
+    ///
+    func displayEmptyFileNameErrorNotice() {
```
I'm curious about when we use this method, since it's not possible to press the `Update` button when it's empty 🤔

partho-maple replied:
You are right. I didn't think like that. Removed.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+
+    /// Displays a Notice onscreen, indicating that the sale price need to be higher than the regular price
+    ///
+    func displayInvalidUrlErrorNotice() {
```
Same question here. When we display this error notice since we disable the `Update` button when the URL is invalid?

partho-maple replied:
You are right. I didn't think like that. Removed.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,356 @@
+        }
+    }
+
+    enum FormType {
```
nit: It will be useful to add a small comment, describing what `FormType` represents. An external developer that doesn't work on this feature, can't understand immediately what are the possibilities around downloadable files.

partho-maple replied:
right. Added

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+import Foundation
+import Yosemite
+
+/// Provides data needed for file settings.
```
nit: We try to be consistent in the descriptions. I saw `downloads`, `downloadable files` and `files`. Let's try to use just `downloadable files`.
```suggestion
/// Provides data needed for downloadable file settings.
```

partho-maple replied:
uprated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+    var fileID: String? { get }
+}
+
+/// Handles actions related to the file settings data.
```
nit
```suggestion
/// Handles actions related to downloadable file settings data.
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+    func hasUnsavedChanges() -> Bool
+}
+
+/// Error cases that could occur in product download file settings.
```
nit:
```suggestion
/// Error cases that could occur in product downloadable file settings.
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+    // Validation
+    private var fileNameIsValid: Bool = false
+    private var fileUrlIsValid: Bool = false
+    private lazy var throttler: Throttler = Throttler(seconds: 0.5)
```
nit: I think it's not used anywhere.

partho-maple replied:
yes. removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+            fileName = file.name
+            fileURL = file.fileURL
+            fileID = file.downloadID
+        }
```
```suggestion
        let file = product.downloadableFiles[safe: downloadFileIndex]
        fileName = file?.name
        fileURL = file?.fileURL
        fileID = file?.downloadID
```

partho-maple replied:
updated accordingly

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        onValidation(isChangesValid(), false)
+    }
```
Nit on code improvements:
```suggestion
        self.fileName = fileName

        let newFileName = fileName
        let oldFileName = product.downloadableFiles[safe: downloadableFileIndex]?.name

        guard newFileName != oldFileName && newFileName?.isEmpty == false else {
            fileNameIsValid = false
            onValidation(fileNameIsValid || fileUrlIsValid, true)
            return
        }

        fileNameIsValid = true

        onValidation(isChangesValid(), false)
    }
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+        }
+
+        onValidation(isChangesValid(), false)
+    }
```
nit:
```suggestion
        self.fileURL = fileURL

        let newURL = self.fileURL
        let oldURL = product.downloadableFiles[safe: downloadableFileIndex]?.fileURL

        guard newURL != oldURL, newURL?.isValidURL() == true && newURL?.isEmpty == false else {
            fileUrlIsValid = false
            onValidation(fileNameIsValid || fileUrlIsValid, true)
            return
        }

        fileUrlIsValid = true

        onValidation(isChangesValid(), false)
    }
```

partho-maple replied:
Updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+            onError(.invalidFileUrl)
+        } else if !fileNameIsValid {
+            onError(.emptyFileName)
+        }
```
It seems that these two cases never happens

partho-maple replied:
removed it altogether

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit FIle/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,184 @@
+            return fileUrlIsValid
+        case .edit:
+            return fileNameIsValid || fileUrlIsValid
+        }
```
In my opinion, it should be better to remove these variables and to hold all the validation checks inside this method.

partho-maple replied:
I also think that. Updated, please check

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift:
```diff
@@ -121,15 +121,17 @@ extension ProductDownloadListViewController {
+        navigationController?.popViewController(animated: true)
     }
 
     @objc private func downloadSettingsButtonTapped() {
```
I think it's clear enought that we are in downloads.
```suggestion
    @objc private func settingsButtonTapped() {
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift:
```diff
@@ -142,21 +144,73 @@ extension ProductDownloadListViewController {
+// MARK: Action - Downloaded Settings
+//
+extension ProductDownloadListViewController {
+    func showDownloadSettings() {
```
```suggestion
    func showDownloadableFilesSettings() {
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift:
```diff
@@ -142,21 +144,73 @@ extension ProductDownloadListViewController {
+                                                                                                    name: fileName ?? "",
+                                                                                                    fileURL: fileURL ?? ""))))
+        }
+        viewModel.completeUpdating(onCompletion: onCompletion)
```
on `completeUpdating` there is a `TODO` that maybe should be handled in this PR.

partho-maple replied:
Yes, handled.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift:
```diff
@@ -142,21 +144,73 @@ extension ProductDownloadListViewController {
+    }
+}
+
+// MARK: Action - Downloaded Settings
```
Downloaded settings? 🤔

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -1153,9 +1153,6 @@ private extension ProductFormViewController {
     }
 
     func onAddEditDownloadsCompletion(data: ProductDownloadsEditableData) {
-        defer {
-            navigationController?.popViewController(animated: true)
-        }
```
I think that we should keep this and remove the `popViewController` inside the `ProductDownloadListViewController` where is not needed.

partho-maple replied:
OK. Updated

pmusolino commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Downloadable Files/ProductDownloadFileViewModelTests.swift:
```diff
@@ -0,0 +1,297 @@
+@testable import WooCommerce
+@testable import Yosemite
+
+final class ProductDownloadFileViewModelTests: XCTestCase {
```
Can we conform all these class naming to `DownloadableFile`, like `ProductDownloadableFileViewModelTests`?

partho-maple replied:
Sorry @pmusolino , I actually didn't understand what you have suggested here. Can you please elaborate?

pmusolino replied:
We are using different class naming for Downloadable Files, so I think that we should update the name of these classes, for example changing  `ProductDownloadFileViewModelTests` to `ProductDownloadableFileViewModelTests` and so on.

partho-maple replied:
yes. I have changed all occurrences of `DownloadFile` and `Downloads` to `DownloadableFile`

pmusolino commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Downloadable Files/ProductDownloadFileViewModelTests.swift:
```diff
@@ -0,0 +1,297 @@
+@testable import WooCommerce
+@testable import Yosemite
+
+final class ProductDownloadFileViewModelTests: XCTestCase {
```
Anyway, these tests look good to me, but probably they need to be updated after the previous changes.

partho-maple replied:
right. updated

pmusolino commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewModel+ChangesTests.swift:
```diff
@@ -271,4 +271,58 @@ final class ProductFormViewModel_ChangesTests: XCTestCase {
+                                             isEditProductsRelease5Enabled: false)
+
+        // Action
+        viewModel.updateDownloadableFiles(downloadableFiles: MockProduct().sampleDownloads(), downloadLimit: 5, downloadExpiry: 1)
```
nit: you shouldn't need to edit the `downloadExpire` here.

partho-maple replied:
I am not actually editing `downloadExpire ` I am only editing `downloadLimit`.  Because in. `MockProduct().product(downloadable: true)`, the value for both `downloadExpire` and `downloadLimit` are `1`. SO at this line I am only changing `downloadLimit` to `5`. And since both value are non optional so, I must pass the original value here which is `1`

pmusolino commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewModel+ChangesTests.swift:
```diff
@@ -271,4 +271,58 @@ final class ProductFormViewModel_ChangesTests: XCTestCase {
+                                             isEditProductsRelease5Enabled: false)
+
+        // Action
+        viewModel.updateDownloadableFiles(downloadableFiles: MockProduct().sampleDownloads(), downloadLimit: 1, downloadExpiry: 5)
```
nit: you shouldn't need to edit the `downloadLimit` here.

partho-maple replied:
Same reason as above.

pmusolino reviewed (changes requested):
Hi @partho-maple and thanks for opening this PR! I've to say that your PRs are still too big. We usually admit PRs with a maximum of 500 lines of code, but you doubled this limit more than one time. They are really hard to review because it's hard to not miss anything. You should split them in small PR, where every single piece of code is easily testable. 👍 

In this PR I left some comments, but I also discovered some issues:
1) When I add a new downloadable file, the bar button items are wrong. it should be just a `Done` button. See #2759 (but it's probably not handled in this PR).
2) After the selection, the tableview cell of a downloadable cell remains in the selected status:
<img src="https://user-images.githubusercontent.com/495617/94424556-b5137d00-018a-11eb-9159-242b6833e151.gif" width=300 />
3) If I try to add a new downloadable file that contains an URL without a path, like `https://www.woocommerce.com`, and if I don't add a File Name, the file will be added locally, but not remotely. So when I press the button for updating the product, the file is not published remotely. What to do in this specific case @Garance91540? Maybe we always should maintain the file name mandatory?

partho-maple committed:
typo fixed and tableviewcell selection behaviors updated

partho-maple committed:
removed default value from downloadableFileIndex and made it optional

partho-maple committed:
unused code cleanup

partho-maple committed:
pr suggestion resolution

partho-maple committed:
refactor file update validation logic

partho-maple committed:
validation logic refactored

partho-maple committed:
PR comments resolution

partho-maple committed:
test case update for ProductDownloadFileViewModelTests

partho-maple committed:
form validation logic updated

partho-maple commented:
> Hi @partho-maple and thanks for opening this PR! I've to say that your PRs are still too big. We usually admit PRs with a maximum of 500 lines of code, but you doubled this limit more than one time. They are really hard to review because it's hard to not miss anything. You should split them in small PR, where every single piece of code is easily testable. 👍
> 
> In this PR I left some comments, but I also discovered some issues:
> 
> 1. When I add a new downloadable file, the bar button items are wrong. it should be just a `Done` button. See #2759 (but it's probably not handled in this PR).
> 2. After the selection, the tableview cell of a downloadable cell remains in the selected status:
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/495617/94424556-b5137d00-018a-11eb-9159-242b6833e151.gif">
> 
> 3) If I try to add a new downloadable file that contains an URL without a path, like `https://www.woocommerce.com`, and if I don't add a File Name, the file will be added locally, but not remotely. So when I press the button for updating the product, the file is not published remotely. What to do in this specific case @Garance91540? Maybe we always should maintain the file name mandatory?

Hi @pmusolino , thanks for the review. I am sorry for going above the line limit. I promise all of my next PRs would be within the line limit.

1. Regarding the bar button item issue, don't worry, I have got it covered with #2759 
2. Sorry. I missed that. Updated accordingly
3. Regarding downloadable file form validation, I have found a post here that has been already implemented on androud counterpart: [https://tinyurl.com/y6em2rzf](https://tinyurl.com/y6em2rzf). So I have updated the logic according to that. Please check if that's OK

partho-maple committed:
broken test case fix

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,290 @@
+            let button = UIBarButtonItem(image: .moreImage,
+                                         style: .plain,
+                                         target: self,
+                                         action: #selector(deleteDownloadableFile))
```
The action called here is wrong, because we should call a method that opens the action sheet. The button inside the action sheet will handle the downloadable file deletion.

partho-maple replied:
Updated to `showBottomActionSheet`

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+                                                           comment: "Placeholder of the cell text field in Product Downloadable File URL")
+        static let urlFooter = NSLocalizedString("This is the url of the file which customers will get accessed to. URLs entered should already be encoded.",
+                                                      comment: "Footer text for Downloadable File URL")
+        static let updateBarButtonAccessibilityIdentifier = "downloadable-file-update-bar-button-item-accessibility-identifier"
```
Since these are the strings related to Localized String, this shouldn't be inside this `.enum`

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,290 @@
+    }
+
+    func enableDoneButton(_ enabled: Bool) {
+        navigationItem.rightBarButtonItems?.forEach({ (barButtonItem) in
+            if barButtonItem.accessibilityIdentifier == ProductDownloadFileViewModel.Strings.updateBarButtonAccessibilityIdentifier {
+                barButtonItem.isEnabled = enabled
+            }
+        })
```
This is a bad pattern, and we should avoid it. I think that the best way to do this is to declare a property for every nav bar button, allocated during the configuration of the bar button items.

partho-maple replied:
updated accordingly

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewController.swift:
```diff
@@ -0,0 +1,290 @@
+                                         action: #selector(completeUpdating))
+            button.accessibilityLabel = NSLocalizedString("Update products' downloadable files' info remotely",
+                                                          comment: "Accessibility label to update products' downloadable files' info remotely")
+            button.accessibilityIdentifier = ProductDownloadFileViewModel.Strings.updateBarButtonAccessibilityIdentifier
```
We should avoid this 👍 (more below)

partho-maple replied:
removed

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+        var fileNameChanged = false
+
+        guard var newFileURL = self.fileURL,
+            newFileURL.isEmpty == false,
```
```suggestion
            newFileURL.isNotEmpty,
```

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+            newFileURL.isValidURL() == true else {
+                return false
+        }
+        newFileURL = newFileURL.addHTTPSSchemeIfNecessary()
```
I'm sorry but it's not clear to me why you are adding the https scheme here (altering the original URL).

partho-maple replied:
As per the last suggestion here: https://tinyurl.com/y6em2rzf,
It says `URL has to be a valid URL, if the user doesn’t enter a protocol, we’ll add “http” implicitly.`
That's why I made that change. But honestly, I think we shouldn't do it. So removing it. Please advise if I should add back it.

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+        var fileUrlChanged = false
+        var fileNameChanged = false
+
+        guard var newFileURL = self.fileURL,
```
`self` is not needed 👍

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+        switch formType {
+        case .add:
+            fileUrlChanged = true
+            if let newFileName = self.fileName,
```
`self` is not needed

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+            }
+        }
+
+        guard let urlLastPathComponent = URL(string: newFileURL)?.lastPathComponent,
+            urlLastPathComponent.isEmpty == false, urlLastPathComponent.split(separator: ".").count >= 2 else {
+                if self.fileName?.isEmpty == false {
+                    switch formType {
+                    case .add:
+                        return fileNameChanged
+                    case .edit:
+                        return fileUrlChanged || fileNameChanged
+                    }
+                } else {
+                    return false
+                }
+        }
```
This looks good, but:
- `self` is not needed
- I imagine a method like this in a URL extension (well unit tested)
- A comment on what are you checking would be useful.

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/Add or Edit File/ProductDownloadFileViewModel.swift:
```diff
@@ -0,0 +1,189 @@
+// MARK: - Convenience Methods
+//
+private extension ProductDownloadFileViewModel {
+    func isChangesValid() -> Bool {
```
I found this method pretty confusing. You check more and more time for the same thing on the same variable, for example you check two times that the file name is not null, that is not empty, and so on. I'd say to check only one time if the file name changed, only one time if the URL changed, only one time if the file index changed.

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewController.swift:
```diff
@@ -142,21 +143,67 @@ extension ProductDownloadListViewController {
+// MARK: Action - Downloadable file settings
+//
+extension ProductDownloadListViewController {
+    func showDownloadableFilesSettings() {
 
+    }
```
nit: add a `TODO`

partho-maple replied:
updated

pmusolino commented on WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift:
```diff
@@ -1146,20 +1146,19 @@ private extension ProductFormViewController {
 
-        let downloadFileListViewController = ProductDownloadListViewController(product: product) { [weak self] data in
-            self?.onAddEditDownloadsCompletion(data: data)
+        let downloadFileListViewController = ProductDownloadListViewController(product: product) { [weak self] (data, hasUnsavedChanges) in
```
It's not clear to me why the `ProductDownloadListViewController` is returning if there are unsaved changes. If there are unsaved changes, you should just compare the original data and the new data returned.

partho-maple replied:
Updated

pmusolino reviewed (changes requested):
I did another review @partho-maple, but I found a regression. In fact, seems that reopening a downloadable file that was edited before does not reflect updated data in the downloadable file screen. The gif below better explain the behavior
<img src="https://user-images.githubusercontent.com/495617/94721615-716d6e80-0356-11eb-8f63-f2b81795ace1.gif" width=300 />


> Regarding downloadable file form validation, I have found a post here that has been already implemented on androud counterpart: https://tinyurl.com/y6em2rzf. So I have updated the logic according to that. Please check if that's OK

The new logic looks good, but I left some comments on it 👍
